Add export traces to gcs

gafda opened the pull request:
This pull request adds support for exporting traces, in addition to logs, to Google Cloud Storage in the `googlecloudstorageexporter`. It updates the configuration, implementation, and tests to handle traces, and improves how encoding extensions are loaded for both logs and traces.

**Traces Export Support:**

* Added support for exporting traces alongside logs, including implementation of `ConsumeTraces`, trace marshaling, and registration of the traces exporter. [[1]](diffhunk://#diff-0d65f863d969afa55f019e94352a68582fc4481604ae72d25cefa3133eedb52eR19-R36) [[2]](diffhunk://#diff-0d65f863d969afa55f019e94352a68582fc4481604ae72d25cefa3133eedb52eL86-L91) [[3]](diffhunk://#diff-0d65f863d969afa55f019e94352a68582fc4481604ae72d25cefa3133eedb52eR163-R174) [[4]](diffhunk://#diff-3602e3108195892e82dc9c6337c257c6b6458f7cefb6b2e0a9442b5e73178f79R20-R30) [[5]](diffhunk://#diff-444365022720b768b6bd00769a57493a121610e137a4ba2d79e2fd6415d094fdR16)

**Configuration and Documentation Updates:**

* Updated the `README.md` and `metadata.yaml` to reflect support for traces and clarify how the `encoding` option works for both logs and traces. [[1]](diffhunk://#diff-a290e38e28b3bbd59d2338455a396b970d85b96c979cfb56ac02be5453eef2caL6-R6) [[2]](diffhunk://#diff-a290e38e28b3bbd59d2338455a396b970d85b96c979cfb56ac02be5453eef2caL21-R22) [[3]](diffhunk://#diff-1cf2732495067d6741fe25706c59e731639948a990e1b26997c9872024522ae5L7-R7)

**Configuration Handling Improvements:**

* Changed the `Encoding` field in the config struct from a pointer to a value, simplifying usage and reducing nil checks. [[1]](diffhunk://#diff-259a21c2d0692b7a8ee971f3bb61b813717f1b034bbb79d64d511adcb5400f4fL14-R14) [[2]](diffhunk://#diff-1715bdd7838e69309f6da74fb5462bc8b817e736cafa23ab68e3ab50b5605d74L31-R31)

**Testing Enhancements:**

* Added and updated tests to cover traces export, including new mock marshalers for traces and both signals, and additional test cases for encoding extension scenarios. [[1]](diffhunk://#diff-a39c72d31c5eb9fef9aa4af4db2570b342af0767c52de777954da6a631fc447bR106-R111) [[2]](diffhunk://#diff-a39c72d31c5eb9fef9aa4af4db2570b342af0767c52de777954da6a631fc447bR126-R153) [[3]](diffhunk://#diff-a39c72d31c5eb9fef9aa4af4db2570b342af0767c52de777954da6a631fc447bL169-R190) [[4]](diffhunk://#diff-a39c72d31c5eb9fef9aa4af4db2570b342af0767c52de777954da6a631fc447bL200-R221) [[5]](diffhunk://#diff-a39c72d31c5eb9fef9aa4af4db2570b342af0767c52de777954da6a631fc447bR231-R255) [[6]](diffhunk://#diff-a39c72d31c5eb9fef9aa4af4db2570b342af0767c52de777954da6a631fc447bR328-R354)

**Configuration Example:**

```yaml
exporters:
  # ...
  googlecloudstorage/traces:
    # ...
    bucket:
      name: "YOUR_BUCKET_NAME"
      project_id: "YOUR_GCP_PROJECT_ID"
      region: "YOUR_BUCKET_REGION"
      # ...

  googlecloudstorage/logs:
    # ...
    bucket:
      name: "YOUR_BUCKET_NAME"
      project_id: "YOUR_GCP_PROJECT_ID"
      region: "YOUR_BUCKET_REGION"
      # ...
```

Copilot committed:
Initial plan

Copilot committed:
Add traces export support to googlecloudstorageexporter

Co-authored-by: ctorrao <22594902+ctorrao@users.noreply.github.com>

Copilot committed:
Fix formatting in exporter_test.go

Co-authored-by: ctorrao <22594902+ctorrao@users.noreply.github.com>

gafda committed:
Refactor encoding field in Config struct to remove pointer and update usage in exporter
feat: enhance OpenTelemetry collector configuration with Prometheus support and improved metrics handling

constanca-m commented on exporter/googlecloudstorageexporter/metadata.yaml:
```diff
@@ -4,7 +4,7 @@ status:
   distributions: [contrib]
   stability:
-    alpha: [logs]
+    alpha: [logs, traces]
```
`traces` should start under `development` stability

constanca-m reviewed (commented):
Thanks for the contribution Gil, usually it is best to start with opening an issue instead of opening the PR already, just to make sure we can have agreement before implementation.

The PR looks good in general, just a couple of concerns

gafda committed:
Update Google Cloud Storage Exporter to support traces export and adjust stability levels

gafda commented on exporter/googlecloudstorageexporter/config.go:
```diff
@@ -11,8 +11,8 @@ import (
 )
 
 type Config struct {
-	Encoding *component.ID `mapstructure:"encoding"`
```
Done! Reverted `Encoding` back to `*component.ID`.

gafda committed:
Address PR review comments: signal validation and encoding error handling

gafda committed:
Improve error handling for traces marshaler loading and update test expectations

gafda commented:
@constanca-m, I'm sorry I took too long to fix your review comments. I've been sick for a while.
Hope everything is in order now.

constanca-m reviewed (commented):
Thanks, almost there

gafda committed:
Enhance error handling for marshaler loading in Google Cloud Storage Exporter

gafda commented on exporter/googlecloudstorageexporter/exporter_test.go:
```diff
@@ -158,7 +185,17 @@ func TestStart(t *testing.T) {
 	t.Run("encoding id not a logs marshaler", func(t *testing.T) {
 		err := gcsExporter.Start(t.Context(), mHost)
-		require.ErrorContains(t, err, "is not a logs marshaler")
+		require.ErrorContains(t, err, "is not a")
```
Changed to `require.ErrorIs`. Added error variables `errNotLogsMarshaler` and `errNotTracesMarshaler` for type-safe error checking. The `loadExtension` function now returns these specific errors, and tests use `require.ErrorIs` for stronger validation.

gafda commented on exporter/googlecloudstorageexporter/exporter_test.go:
```diff
@@ -353,6 +390,66 @@ func TestConsumeLogs(t *testing.T) {
+	}
+	id := component.MustNewID(encodingLogsOnlyID)
+	// Create a traces exporter with logs-only encoding
+	exp, err := newStorageExporter(
```
Refactored to use `newTestGCSExporter` with an optional signal parameter (defaults to `signalTypeLogs`). The test structure now mirrors `TestConsumeLogs`. However, the behavior differs slightly: when a logs-only encoding is configured for traces, the exporter gracefully falls back to JSON marshaler with a warning instead of failing. This prevents production failures when users configure global encoding extensions that only support logs (like `text_encoding`). Logs remain strict and throw errors for incompatible encodings.

constanca-m reviewed (approved):
Thanks @gafda, looks good

gafda committed:
Enhance log and trace consumption tests with improved error handling and encoding validation

gafda committed:
Fix linting issues and add changelog entry

- Simplify early-return pattern in traces marshaler loading
- Remove unused parameter from loadExtension function
- Fix struct field formatting in test file
- Add changelog entry for traces export support
